unit-tests/dds/test-metadata-syncer.py

```diff
@@ -0,0 +1,330 @@
+    test.check( last_metadata ) and test.check_equal( md_id( last_metadata ), 1 )
+    test.check_equal( len(dropped_metadata), 0 )
+
+with test.closure( 'Enqueue metadata+image -> out' ):
+    syncer = new_syncer()
+    syncer.enqueue_metadata( 1, new_metadata( 1 ) )
+    test.check_equal( last_image, None )
+    test.check_equal( last_metadata, None )
+    test.check_equal( len(dropped_metadata), 0 )
+    syncer.enqueue_frame( 1, new_image( 1 ) )
+    test.check( last_image ) and test.check_equal( image_id( last_image ), 1 )
+    test.check( last_metadata ) and test.check_equal( md_id( last_metadata ), 1 )
+    test.check_equal( len(dropped_metadata), 0 )
+
+with test.closure( 'Enqueue 1 image after some metadata -> drops + match' ):
+    syncer = new_syncer()
+    for i in range(4):
+        syncer.enqueue_metadata( i, new_metadata( i ) )
+    test.check_equal( last_image, None )
+    test.check_equal( last_metadata, None )
+    test.check_equal( len(dropped_metadata), 0 )
+    syncer.enqueue_frame( 2, new_image( 2 ) )
+    test.check( last_image ) and test.check_equal( image_id( last_image ), 2 )
+    test.check( last_metadata ) and test.check_equal( md_id( last_metadata ), 2 )
+    test.check_equal( len(dropped_metadata), 2 )  # 0 and 1
+
+with test.closure( 'Enqueue 1 image then earlier metadata -> nothing out; metadata is dropped' ):
+    syncer = new_syncer()
+    syncer.enqueue_frame( 1, new_image( 1 ) )
+    test.check_equal( last_image, None )
+    test.check_equal( last_metadata, None )
+    test.check_equal( len(dropped_metadata), 0 )
+    syncer.enqueue_metadata( 0, new_metadata( 0 ) )
+    test.check_equal( last_image, None )
+    test.check_equal( last_metadata, None )
+    test.check_equal( len(dropped_metadata), 1 )  # not going to be any frame for it
+
+with test.closure( 'Enqueue 1 image then later metadata -> image out' ):
+    syncer = new_syncer()
+    syncer.enqueue_frame( 1, new_image( 1 ) )
+    test.check_equal( last_image, None )
+    test.check_equal( last_metadata, None )
+    test.check_equal( len(dropped_metadata), 0 )
+    syncer.enqueue_metadata( 2, new_metadata( 2 ) )
+    test.check( last_image ) and test.check_equal( image_id( last_image ), 1 )
+    test.check_equal( last_metadata, None )
+    test.check_equal( len(dropped_metadata), 0 )
+
+with test.closure( 'Enqueue 2 images then matching metadata for the second -> both out' ):
+    syncer = new_syncer()
+    syncer.enqueue_frame( 1, new_image( 1 ) )
+    test.check_equal( last_image, None )
+    test.check_equal( last_metadata, None )
+    test.check_equal( len(dropped_metadata), 0 )
+    syncer.enqueue_frame( 2, new_image( 2 ) )
+    test.check_equal( last_image, None )
+    test.check_equal( last_metadata, None )
+    test.check_equal( len(dropped_metadata), 0 )
+    syncer.enqueue_metadata( 2, new_metadata( 2 ) )
+    test.check( last_image ) and test.check_equal( image_id( last_image ), 2 )
```

Comment:
Maybe add a counter in `on_frame_ready` and check that both frames were handled

Reply:
We could record the frames that were received in `on_frame_ready`... I'll see if I can add.